cumulus/pallets/executive/src/lib.rs

```diff
@@ -359,6 +359,12 @@ impl<
 		res
 	}
 
+	/// Variant of [`apply_extrinsic`] to return the storage root after applying the extrinsic.
+	pub fn apply_extrinsic_with_post_state_root(uxt: Block::Extrinsic) -> Vec<u8> {
+		let _ = Self::apply_extrinsic(uxt);
+		Self::storage_root()
+	}
```

Comment:
Not sure why we need this method since we can call both `Self::apply_extrinsic()` and `Self::storage_root()` using public API already

Reply:
Because you can't call them while checking the execution proof and need this API to re-execute the extrinsic precisely, https://github.com/subspace/subspace/blob/7a8f22df55/cumulus/client/cirrus-executor/src/tests.rs#L248-L257

Reply:
If I can call the method that calls both, I surely can call them separately.

Specifically, the only reason it is here is to be called from the runtime API implementation, so we could have called these from runtime API directly without introducing a new method, it would be even easier to understand what is happening (to me at least).

Reply:
>If I can call the method that calls both, I surely can call them separately.

It's true for the regular case, but not for the case of fraud proof in which you don't call these from runtime API but spin up a new VM and feed in the method, call data accordingly.

Reply:
But you still call runtime API in that case, not internal pallet method.

Reply:
Nope. The context of calling a runtime API and proving an execution is different, this API is basically for verifying the post state root, see this discussion https://github.com/paritytech/substrate/discussions/10987 .

Reply:
You're calling `SecondaryApi_apply_extrinsic_with_post_state_root`. That is a runtime API. I understand that you're not calling it with standard runtime API abstraction of Substrate, but that doesn't change the fact that it is a runtime API and not a call to the pallet directly. Hence there is no need to add a method to the pallet, you can call 2 methods instead of one in the `SecondaryApi_apply_extrinsic_with_post_state_root`:
https://github.com/subspace/subspace/blob/7a8f22df558f775e18c2dc619f9c71455717cec4/cumulus/parachain-template/runtime/src/lib.rs#L440-L442

Reply:
Calling the runtime API under the fraud proof context has to use the low-level API like this https://github.com/subspace/subspace/blob/7a8f22df55/cumulus/client/cirrus-executor/src/tests.rs#L248-L261, where you have to prepare everything for calling that API(state, code, method, call data) why you bother calling these separately and not using a convenient API?

Reply:
That is what I'm saying, it is calling `SecondaryApi_apply_extrinsic_with_post_state_root`, which is the code I linked above. Why code above can't be this?:
```rust
 fn apply_extrinsic_with_post_state_root(extrinsic: <Block as BlockT>::Extrinsic) -> Vec<u8> { 
 	let _ = Executive::apply_extrinsic(extrinsic);
 	Executive::storage_root() 
 }

Reply:
Okay, I see you were saying, you meant why not put it on the runtime level and put it inside the executive pallet. Well, I don't have a strong opinion, both is fine to me, putting it on the runtime level just needs to expose `storage_root` in the executive pallet.

Reply:
It already is exposed and public, it is part of `Executive` trait.

Reply:
It's not yet public. https://github.com/subspace/subspace/blob/33c2a30eacbfc32ae8f35cda3e34575d805d401a/cumulus/pallets/executive/src/lib.rs#L187

Reply:
This should be what you meant https://github.com/subspace/subspace/pull/286/commits/7c524dc0f9a93a1b762168a868628ac0a0e68cae. BTW, there is no `Executive` trait :D.

Reply:
Hm, indeed, I read it wrong